test/validation/api/comp/comp.c

```diff
@@ -0,0 +1,589 @@
+	    !(comp_capa.hash_algo.bit.none))
+		rc = -1;
+	if (hash_alg == ODP_COMP_HASH_ALG_SHA1 &&
+	    !(comp_capa.hash_algo.bit.sha1))
+		rc = -1;
+	if (hash_alg == ODP_COMP_HASH_ALG_SHA256 &&
+	    !(comp_capa.hash_algo.bit.sha256))
+		rc = -1;
+
+	CU_ASSERT(!rc);
+
+	/* Create a compression session */
+	odp_comp_session_param_init(&ses_params);
+	ses_params.op = op;
+	ses_params.comp_algo = comp_alg;
+	ses_params.hash_algo = hash_alg;
+	ses_params.compl_queue = suite_context.queue;
+	ses_params.mode = suite_context.op_mode;
+
+	session = odp_comp_session_create(&ses_params);
+	CU_ASSERT_FATAL(session != ODP_COMP_SESSION_INVALID);
+	CU_ASSERT(odp_comp_session_to_u64(session) !=
+		odp_comp_session_to_u64(ODP_COMP_SESSION_INVALID));
+
+	/* Allocate compression output packet */
+	outpkt = odp_packet_alloc(suite_context.pool, outtext_len);
+	CU_ASSERT(outpkt != ODP_PACKET_INVALID);
+
+	op_params.out_data_range.offset = 0;
+	op_params.out_data_range.length = outtext_len;
+	op_params.in_data_range.offset = 0;
+	op_params.in_data_range.length = odp_packet_len(inpkt);
+	op_params.session = session;
+
+	if (suite_context.op_mode == ODP_COMP_OP_MODE_SYNC) {
+		rc = odp_comp_op(&inpkt, &outpkt, 1, &op_params);
+		CU_ASSERT(rc >= 0);
+		if (rc < 0)
+			goto cleanup;
+	} else {
+		odp_event_t event;
+		odp_packet_t packet;
+
+		rc = odp_comp_op_enq(&inpkt, &outpkt, 1, &op_params);
+		CU_ASSERT(rc == 1);
+		if (rc <= 0)
+			goto cleanup;
+		/* Poll completion queue for results */
+		do {
+			event = odp_queue_deq(suite_context.queue);
+		} while (event == ODP_EVENT_INVALID);
+		CU_ASSERT(ODP_EVENT_PACKET == odp_event_type(event));
+		CU_ASSERT(ODP_EVENT_PACKET_COMP ==
+			  odp_event_subtype(event));
+
+		packet = odp_comp_packet_from_event(event);
+		CU_ASSERT(packet != ODP_PACKET_INVALID);
+		CU_ASSERT(packet == outpkt);
+	}
+
+	rc = odp_comp_result(&comp_result, outpkt);
+	CU_ASSERT(!rc);
+	CU_ASSERT(comp_result.status == ODP_COMP_STATUS_SUCCESS);
+	CU_ASSERT(comp_result.output_data_range.offset == 0);
+	odp_packet_trunc_tail(&outpkt,
+			      odp_packet_len(outpkt) -
+			      comp_result.output_data_range.length,
+			      NULL, NULL);
+
+cleanup:
+
+	rc = odp_comp_session_destroy(session);
+	CU_ASSERT(!rc);
+
+	if (rc < 0) {
+		odp_packet_free(outpkt);
+		return ODP_PACKET_INVALID;
+	}
+
+	return outpkt;
+}
+
+static void packet_cmp(odp_packet_t pkt,
+		       const uint8_t *text,
+		       unsigned int text_len)
+{
+	odp_packet_seg_t seg;
+	uint32_t cmp_offset = 0, outlen = 0;
+	uint32_t compare_len = 0;
+	uint8_t *outdata;
+
+	seg = odp_packet_first_seg(pkt);
+	do {
+		outdata = odp_packet_seg_data(pkt, seg);
+		outlen = odp_packet_seg_data_len(pkt, seg);
+		compare_len = outlen < (text_len - cmp_offset) ?
+			outlen : (text_len - cmp_offset);
+
+		CU_ASSERT(!memcmp(outdata,
+				  text + cmp_offset, compare_len));
+		cmp_offset += compare_len;
+		seg = odp_packet_next_seg(pkt, seg);
+	} while (seg != ODP_PACKET_SEG_INVALID || cmp_offset < text_len);
```

Comment:
This doesn't seem right. If `seg == ODP_PACKET_SEG_INVALID` it doesn't matter if `cmp_offset < text_len`. You can't reference offsets into `ODP_PACKET_SEG_INVALID`.